### PR TITLE
johanna/3850 multiplex results form behavior bugs

### LINKED ISSUE #3850: [BUG] Multiplex Conduct Tests Screen: Radio button for COVID & Flu B Results doesn't take on the first try
### Description
On the "Conduct tests" screen, when viewing a Test Card for a Multiplex test -- when clicking the radio buttons to select "positive" or "negative" for COVID, Flu A, and Flu B - if you first select the "Negative" radio button for Flu B, then attempt to select "Positive" - the radio button for Negative stays selected. Clicking "Positive" under Flu B a second time finally switches the result.

### Steps to reproduce

1. Go to the Conduct Test screen while a test is currently underway.
2. Click on the radio buttons to select positive for COVID, positive for Flu A, and Negative for Flu B
3. Attempt to switch the radio button result for Flu B from negative to positive.
4. The result does not immediately change to positive, until clicking a second time.
5. The same happens if you attempt to go from positive to negative. (It requires a second click.)
6. This bug is not present for Flu A. Only COVID and Flu B.

### Expected behavior
When selecting a radio button, the action should change immediately. 
The user should not be required to click a second time.
This has the potential to lead to incorrect test results being submitted.

### Additional context
Bug experienced on a MacBookPro running macOS Monterey Version 12.4
Chrome Version 102.0.5005.61 (Official Build) (x86_64)
Screenshot is attached.
radio-buttons-do-not-switch-as-expected

### PR DESCRIPTION
**Pull request by johanna-skylight**
# FRONTEND PULL REQUEST

## Related Issue

Fixes #3850 , #3840 , #3851 , #3839 

## Changes Proposed

This PR was branch from Nick Clydes PR. The changes that I introduced are:
- General fixes for styling issues.
- The Test Result form has been renamed to CovidResultForm and a new MultiplexResultForm has been introduced.
- The QueueItem component is a smart component that decides what type of result will be asked in the test card (based on the device selected) and renders the correct result form for it.
- The result forms are dumb components that only care to render the correct UI and perform a lightweight validation to enable and disable the submit button. Once the user makes changes to the result these dumb components notify the smart component Queue Item of the change and wait for it to make the necessary API calls and return back the new persisted state of the result. The dumb components do not hold any result inner state, they are display only.

## Testing

Start a multiplex test and check below:
- When unchecking current date/time to enter custom date the date fields do not move below.
- Check the multiplex info tooltip next to the submit button displays the bubble in a way that the text can be read.
- Check that when checking the results radio buttons and inconclusive check boxes the clicks are properly processed in the UI and that the validation is correct for disabling and enabling the submit button.
 
## Screenshots / Demos

- For large changes, please pair with a designer to ensure changes are as intended

### Design
- [ ] Any UI/UX changes have a designer as a reviewer, and changes have been approved
- [X] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke-tested by both the engineering and design teams

### CODE REVIEW

**Comment by johanna-skylight** [comment]
I am still working on the e2e testing failure but I think this PR is good for a review.

Reminder most of the backend code was branch off Nick's PR and has not been modified.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java [hunk 1]
```diff
@@ -265,27 +264,40 @@ public TestOrder editQueueItem(
...
   @AuthorizationConfiguration.RequirePermissionUpdateTestForTestOrder
   public TestOrder editQueueItemMultiplex(
       UUID testOrderId, UUID deviceSpecimenTypeId, List<DiseaseResult> results, Date dateTested) {
-    TestOrder savedOrder =
-        editQueueItem(
-            testOrderId,
-            deviceSpecimenTypeId,
-            getCovidResultFromMultiplexList(results).toString(),
-            dateTested);
+    lockOrder(testOrderId);
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:276]: Just a note that I edited these functions to be a pretty direct copy of the original `editQueueItem` and `addTestResult`. Nick had originally referenced those functions and just used these new functions to update the non-covid test results, but with the corrections and removals logic added it became much easier to simply duplicate the entire function and update the Results in a single place. After these have been in place for a while we can remove the old methods.

File: frontend/src/app/testQueue/QueueItem.tsx [hunk 2]
```diff
@@ -239,13 +200,13 @@ const QueueItem = ({
   };
 
   const [mutationError, updateMutationError] = useState(null);
-  const [removePatientFromQueue] = useMutation(REMOVE_PATIENT_FROM_QUEUE);
-  const [submitTestResult, { loading }] = useMutation(SUBMIT_TEST_RESULT);
+  const [removePatientFromQueue] = useRemovePatientFromQueueMutation();
```
Comment (emmastephenson) [inline frontend/src/app/testQueue/QueueItem.tsx:203]: Thank you for editing these other mutations to be in the new style 🙏

File: frontend/src/app/testQueue/QueueItem.tsx [hunk 3]
```diff
@@ -440,28 +433,25 @@ const QueueItem = ({
...
-            response.data.editQueueItem.deviceSpecimenType.deviceType
-              .testLength as number
+            response?.data?.editQueueItemMultiplex?.deviceSpecimenType
+              ?.deviceType.testLength as number
           );
           updateDeviceTestLength(
-            response.data.editQueueItem.deviceSpecimenType.deviceType
-              .testLength as number
+            (response?.data?.editQueueItemMultiplex?.deviceSpecimenType
+              ?.deviceType?.testLength as number) ?? 0
```
Comment (emmastephenson) [inline frontend/src/app/testQueue/QueueItem.tsx:454]: Is this the default if no `testLength` is found? If so I would update to 15 instead of 0, so the timer continues to work.

File: frontend/src/app/testQueue/TestQueue.test.tsx [hunk 4]
```diff
@@ -10,8 +10,13 @@ import { MemoryRouter } from "react-router-dom";
...
 
-import TestQueue, { queueQuery } from "./TestQueue";
-import { REMOVE_PATIENT_FROM_QUEUE } from "./QueueItem";
+import {
+  GetFacilityQueueMultiplexDocument,
+  RemovePatientFromQueueDocument,
+} from "../../generated/graphql";
+
+import TestQueue from "./TestQueue";
+// import { useRemovePatientFromQueueDocyment } from "./QueueItem";
```
Comment (emmastephenson) [inline frontend/src/app/testQueue/TestQueue.test.tsx:19]: This import statement can probably be removed

File: frontend/src/app/testResults/MultiplexResultInputForm.test.tsx [hunk 5]
```diff
@@ -0,0 +1,254 @@
+import React from "react";
+import { render, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+
+import MultiplexResultInputForm from "./MultiplexResultInputForm";
+
+jest.mock("uuid");
```
Comment (emmastephenson) [inline frontend/src/app/testResults/MultiplexResultInputForm.test.tsx:7]: Out of curiosity - what does mocking the UUID enable?

**Review by emmastephenson — COMMENTED** [review]
Overall it looks great, thanks for all your extra work on this Johanna! I'm only holding off on approval until I get the chance to test it in `dev` tomorrow.

File: frontend/src/app/testQueue/QueueItem.tsx (hunk 3 above)
Comment (johanna-skylight) [reply frontend/src/app/testQueue/QueueItem.tsx:454]: yeah, the linter was giving me errors because of a possible null/undefined scenario. I will update this to 15. Thanks for catching this!

File: frontend/src/app/testResults/MultiplexResultInputForm.test.tsx (hunk 5 above)
Comment (johanna-skylight) [reply frontend/src/app/testResults/MultiplexResultInputForm.test.tsx]: To be honest it ended up there because I copied the structure of this test from the old testResultForm test file. But I checked and is not doing anything so I removed it.

File: frontend/src/app/testResults/MultiplexResultInputForm.tsx [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            name={`flu-b-test-result-${queueItemId}`}
+            selectedRadio={resultsMultiplexFormat.fluB}
+            wrapperClassName="prime-radio__group"
+            disabled={isSubmitDisabled}
+          />
+        </div>
+      </div>
+      <div className="prime-test-result-submit grid-row">
+        <div className="grid-col-fill">
+          <Checkboxes
```
Comment (emyl3) [inline frontend/src/app/testResults/MultiplexResultInputForm.tsx:246]: Noticed some strange behavior with this checkbox unlike other checkboxes. When you click "Inconclusive" and try to uncheck it, it never unchecks. I created a screen recording. Let me know if this is intended behavior.

https://user-images.githubusercontent.com/20211771/175313952-ee63bd71-95c3-4df6-9768-3fa88bb8c7d9.mov

Comment (johanna-skylight) [reply frontend/src/app/testResults/MultiplexResultInputForm.tsx:246]: That is a weird behavior, but it is not a bug. What is happening is that the backend does not support setting back the results to a null state. So basically when you are trying to uncheck the box you are setting all the results back to null (backend only accepts positive, negative, undetermined mutations) so the request kinda gets ignored. 

We should talk about this today in eng standup @emmastephenson

File: frontend/src/app/testQueue/QueueItem.tsx (hunk 2 above)
Comment (emyl3) [reply frontend/src/app/testQueue/QueueItem.tsx:203]: Oh thanks for doing this, Johanna! and also pointing this out, Emma! will be sure to do this in my PRs too 😅

File: frontend/src/app/testResults/MultiplexResultInputForm.tsx (hunk 6 above)
Comment (emyl3) [reply frontend/src/app/testResults/MultiplexResultInputForm.tsx:246]: Oh I see. Thank you for letting me know the cause of this issue, Johanna! 

If it helps with the discussion, the behavior for correcting an inconclusive test looks like this 😓 :

https://user-images.githubusercontent.com/20211771/175322654-12211847-ea15-4b05-b8dc-7ac73422955e.mov

Comment (johanna-skylight) [reply frontend/src/app/testResults/MultiplexResultInputForm.tsx:246]: that does look like a bug, will check into it

Comment (johanna-skylight) [reply frontend/src/app/testResults/MultiplexResultInputForm.tsx:246]: It seems that after the results are corrected all the possible values get returned back to the frontend. I am not sure if this is the behavior that is expected.
Screen Shot 2022-06-23 at 10 32 35 AM

Then depending on what is read first on the results array the front end displays what is told.

Comment (emyl3) [reply frontend/src/app/testResults/MultiplexResultInputForm.tsx:246]: another weird edge case is if a user is in the middle of correcting an inconclusive test and go to mark a result for another test in the queue 🧐 

https://user-images.githubusercontent.com/20211771/175327552-d0773797-6672-4737-8057-6421ef700f9a.mov

File: frontend/src/app/testQueue/TestQueue.tsx [hunk 7]
```diff
@@ -43,7 +44,7 @@ const emptyQueueMessage = (
 );
 
 export const queueQuery = gql`
-  query GetFacilityQueue($facilityId: ID!) {
```
[the remark this answers is not in the document]
Comment (johanna-skylight) [reply frontend/src/app/testQueue/TestQueue.tsx:46]: I didn't want to mess with this file too much because I only touch it for some css styling issues. Better to leave this migration to be bundle with a ticket that allows to throughly test this component.

**Comment by emmastephenson** [comment]
@johanna-skylight I just pushed the latest changes to `dev` after adding a fix for the corrections issue, and I think we may need to increase the debounce time again - with all three results, it's possible to click too quickly between them and send a bunch of concurrent editQueueItem mutations.

This video hopefully demonstrates what I mean:

https://user-images.githubusercontent.com/80282552/175423939-eec2c3d4-b3c1-44d6-82cf-65d8e60d8a5d.mp4

File: frontend/cypress/integration/03-conduct_test_spec.js [hunk 8]
```diff
@@ -34,7 +34,7 @@ describe("Conducting a test", () => {
     });
     cy.get(".prime-home").contains(patientName);
     queueCard = "div.prime-queue-item:last-of-type";
-    cy.get(queueCard).contains("SARS-CoV-2 results");
+    cy.get(queueCard).contains("COVID-19 results");
```
Comment (johanna-skylight) [inline frontend/cypress/integration/03-conduct_test_spec.js:37]: This is the e2e test that was messing up with me like for two days! T_T

**Review by kenieh — APPROVED** [review]
LGTM! Thank you for all your work on this @johanna-skylight

File: frontend/src/app/testResults/MultiplexResultInputForm.test.tsx [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
-import { render, screen } from "@testing-library/react";
+import { fireEvent, render, screen } from "@testing-library/react";
 import userEvent from "@testing-library/user-event";
 
 import MultiplexResultInputForm from "./MultiplexResultInputForm";
 
-jest.mock("uuid");
+import spyOn = jest.spyOn;
+
+// jest.mock("uuid");
```
Comment (emmastephenson) [inline frontend/src/app/testResults/MultiplexResultInputForm.test.tsx:9]: I think this can also be removed

**Review by emmastephenson — APPROVED** [review]
LGTM, thanks for all your hard work on this Johanna!!

**Review by emmastephenson — APPROVED** [review]

**Review by emyl3 — APPROVED** [review]
Thank you for working on this! It looks great to me!! 🤩 😹
